### ISSUE TITLE
label leakage datasets

### ISSUE THREAD

**Issue by amueller** [openml/OpenML#474]
Here's an example "bot" that detects label leakage:

```python
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import Imputer

import openml

# get all classification tasks
tasks = openml.tasks.list_tasks(task_type_id=1)

# get all "small" datasets that are used for classification
dids = set()
for tid, task in tasks.items():
    if task['NumberOfFeatures'] * task['NumberOfInstances'] < 10000:
        dids.add(task['did'])

for did in dids:
    try:
        ds = openml.datasets.get_dataset(did)
        X, y = ds.get_data(target=ds.default_target_attribute)

    except:
        print("error")
    X = Imputer().fit_transform(X)
    n_classes = int(float(ds.qualities['NumberOfClasses']))
    if n_classes == -1:
        continue
    acc = DecisionTreeClassifier(max_depth=n_classes - 1).fit(X, y).score(X, y)
    if acc > .99:
        print(did, acc)
```
results:
480 1.0
694 1.0
759 1.0
778 1.0
818 1.0
874 1.0
880 1.0
885 1.0
951 1.0
965 1.0
969 1.0
1121 1.0
1167 1.0
1472 1.0
1482 1.0

**Comment by amueller** [comment]
uuhh bug! whoops. Fixed now.

**Comment by amueller** [comment]
For example in 778 the label is a single threshold on the first axis. This is probably because the dataset is the result of binarizing a regression dataset. ping @joaquinvanschoren

**Comment by amueller** [comment]
In https://www.openml.org/d/1121 the target is simply identical to one of the features (which you can easily see even on the website)

**Comment by amueller** [comment]
for issues like these, should the server check for it or should we have bots check for it? Do we have some way to disable these datasets via the API? I'm not sure what the consensus was.

**Comment by amueller** [comment]
Result when including some bigger datasets:
4550 1.0
478 1.0
480 1.0
694 1.0
728 0.993089832182
759 1.0
778 1.0
818 1.0
874 1.0
880 1.0
885 1.0
951 1.0
965 1.0
969 1.0
1121 1.0
1466 1.0
1525 1.0
1526 1.0
40665 1.0

**Comment by amueller** [comment]
Some of these, like the two-class iris dataset are not really information leakage but just trivial data sets. So I'm not sure if we should just remove all of them, or remove the ones that are bugs and keep the trivial ones and tag them as trivial. For example 1526 might be interesting for someone for some historical reason and they might want to store it on openml? But I'm not sure if that's a strong argument.

**Comment by joaquinvanschoren** [comment]
Thanks! A decision tree accuracy of 100% doesn't really imply label leakage though, especially 
not for highly skewed datasets.
Here's the result when using Cohen's kappa (balanced accuracy or multiclass AUC are not available in sklearn yet).

```
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import Imputer
from sklearn.model_selection import cross_val_score
from sklearn.metrics import cohen_kappa_score, make_scorer

import numpy as np

import openml

# use a score function sensitive to class imbalance
kappa_scorer = make_scorer(cohen_kappa_score)

# get all classification tasks
tasks = openml.tasks.list_tasks(task_type_id=1)

# get all "small" datasets that are used for classification
dids = set()
for tid, task in tasks.items():
    if task['NumberOfFeatures'] * task['NumberOfInstances'] < 10000:
        dids.add(task['did'])

for did in dids:
    try:
        ds = openml.datasets.get_dataset(did)
        X, y = ds.get_data(target=ds.default_target_attribute)
        X = Imputer().fit_transform(X)
        n_classes = int(float(ds.qualities['NumberOfClasses']))
        if n_classes == -1:
            continue
        model = DecisionTreeClassifier(max_leaf_nodes=n_classes).fit(X, y)
        kappa = np.mean(cross_val_score(model, X, y, scoring=kappa_scorer, cv=2))
        if kappa > .99:
            print(did, kappa)
    except:
        print("error: ", did)
```

I'm only getting 4 datasets now:
778 0.992059490799     # trivial binarized dataset
965 1.0  
1121 1.0

778: seems to have become trivial dataset after binarization
965: the binarization has separated the mammals from non-mammals, which is perfectly predicted by the 'milk' feature. Funny. I propose to flag that feature to be ignored.
1121: 'second_char_vowel' directly maps to the class. Propose to ignore that feature.

**Comment by amueller** [comment]
@joaquinvanschoren btw which definition of balanced accuracy do you mean? The unnormalized version is there as ``macro_average_recall`` if that's what you mean.

**Comment by amueller** [comment]
How to the different APIs handle ignored features? And where is the documentation on ignored features?
